### ISSUE TITLE
Epic: How to re-use existing project scripts?

### ISSUE THREAD

**Issue by loujaybee** [gitpod-io/gitpod#7627]
There are lots of situations where projects have existing scripts for building/starting their projects: 

* CI configurations (GitHub actions workflows)
* Makefiles
* devcontainer.json [1]
* Existing Docker configurations

This epic is to collate ideas about how we can leverage these scripts so that customers need to do minimal manual changes to their project in order to setup Gitpod support. Ideally gitpod detects and re-uses as much as possible from the existing repository.

**Comment by csweichel** [comment]
Sounds a lot like the project configuration inference @gitpod-io/engineering-meta just shipped, no?

**Comment by JanKoehnlein** [comment]
See also #7392

**Comment by jankeromnes** [comment]
> Sounds a lot like the project configuration inference [...] just shipped, no?

True. 🎯 I read this like a collection of feature requests (4 bullet points) & discussion about approach (1. vs 2.).

> CI configurations (GitHub actions workflows)

This sounds like a good idea in principle, but I wonder:

a. how easy it is to parse a GitHub Action / other CI config file (they generally come with complicated multi-step workflows, dependencies, extra features like between-step caches, or third-party workflows, etc)
b. how relevant a CI job is to produce an actual dev environment

For example, if we look at two of our repositories (one using GitHub Actions, another using CircleCI), I don't see much that we can use to infer a relevant dev environment:
- https://github.com/open-vsx/publish-extensions/tree/master/.github/workflows
- https://github.com/gitpod-io/workspace-images/blob/master/.circleci/config.yml

> Makefiles

The inferrer actually does something super basic already (i.e. "Is there a Makefile? If yes --> add task `init: make`)

https://github.com/gitpod-io/gitpod/blob/94f3b38615535414f59d6aa997e467efffaad593/components/server/src/config/config-inferrer.ts#L105-L115

We could probably be smarter here, e.g. if we know common Makefile rule names used for development (e.g. `./configure` scripts come to mind, which isn't in Makefiles, but is commonly found next to one).

> devcontainer.json [[1](https://github.com/gitpod-io/roadmap/issues/16)]

(C.f. linked issue)

> Existing Docker configurations

Interesting. I guess we could detect files called `Dockerfile` (and possibly other common, more dev-oriented names, e.g. `dev.Dockerfile` if that's a common thing), then try to use that as a base for the dev environment.

Does this always make sense? Would be cool to research this, e.g. anecdotally with a few example repositories.

> There are two broad options (for the different configuration files)
> 1. Directly integrate with the configurations, like the devcontainer standard [1]
> 2. Map from these configurations to Gitpod configurations (either implicitly behind the scenes, or via explicit suggestion)

My 2 cents are that 2. makes the most sense currently (i.e. generate an explicit `.gitpod.yml` based on what we know, then let users take it from there -- doing implicit things sounds more dangerous, especially since users have no way to fix our inevitable mistakes).

Longer-term, if there is a standard that emerges, I agree that we should play nice and participate in the standardization effort + natively support the new standard early.

**Comment by loujaybee** [comment]
> Sounds a lot like the project configuration inference @gitpod-io/engineering-meta just shipped, no?

Yep—it definitely relates! The issue originates from some usability conversations, and conversations with/in the community. 

As @jankeromnes mentioned, this issue is more on a higher-level than inference for specific config (unless our inference "engine" is somewhat generic / almost plugin-oriented. This issue is more about Gitpod's philosophy on integrating existing scripts. If/when we come to a more concrete decision we can update the issue/community on our stance here 🙏 🚀 

Also, all great points/thoughts @jankeromnes ! 🙏

**Comment by maphew** [comment]
Yes please! I searched for and found this issue because my naive attempt to use _"Existing Docker configurations"_ by simply referencing the project Dockerfile in `.gitpod.yml` did not fare well, and I lack the knowledge so far to rebuild it for Gitpod land. With that in mind I present the Dockerfiles under https://github.com/OSGeo/gdal/tree/master/docker as fodder for you to consider while working on this feature.

**Comment by axonasif** [comment]
Hey @maphew, I think you were trying to use the alpine dockerfiles that you have on that repo, please keep in mind that we don't support alpine images as the base for now.

**Comment by maphew** [comment]
@axonasif  I was using ubuntu. I've since figured how to use the published image and then modify it post boot with RUN commands, so I'm not blocked anymore (tho' I still haven't cracked the nut of getting sudo to work passwordless):
- https://github.com/maphew/gdal/commit/4eb513dde5c5c0348b260f632bc6957dbc81bb00
- https://github.com/maphew/gdal/blob/4eb513dde5c5c0348b260f632bc6957dbc81bb00/.gitpod.yml

Here's my failed attempt to use the dockerfile in place: https://github.com/maphew/gdal/commit/bcf888d4819547531018f9ef16737287b8c9578e
